src/command_helpers.rs

```diff
@@ -307,8 +263,27 @@ pub(crate) fn try_wait_on_child(
     program: &str,
     child: &mut Child,
     stdout: &mut dyn io::Write,
+    child_stderr_reader: &mut Option<BufReader<std::process::ChildStderr>>,
 ) -> Result<Option<()>, Error> {
-    match child.try_wait() {
+    // Check the child status THEN forward stderr messages, so that we don't race
+    // between the child printing messages and then exiting.
+    let wait_result = child.try_wait();
+
+    if let Some(child_stderr_reader) = child_stderr_reader.as_mut() {
```

Comment:
I'm thinking if it makes more sense to do this in the async runtime, i.e. if we add poll/select/epoll/waitfor_multiple_objects in the future to speed it up, then it makes sense to do it there.

However since it'd be done in the future, you don't have to change this code.

Some refactoring to extract a new fn is enough.